File: tests/example-frames/deformers/skin-weights.c
// $dep ufbx skinned_human.fbx
// $ clang ufbx.c main.c -lm -o example
// $ ./example skinned_human.fbx Female > result.obj
#include "ufbx.h"
#include <stdio.h>
#include <stdlib.h>

// -- EXAMPLE_SOURCE --

typedef struct Bone {
    uint32_t node_index;
    ufbx_matrix geometry_to_bone;
} Bone;

// Non-indexed triangulated skinned mesh
typedef struct Mesh {
    Bone *bones;
    size_t num_bones;
    Vertex *vertices;
    size_t num_vertices;
} Mesh;

Mesh process_skinned_mesh(ufbx_mesh *mesh, ufbx_skin_deformer *skin)
{
    size_t num_triangles = mesh->num_triangles;
    Vertex *vertices = calloc(num_triangles * 3, sizeof(Vertex));
    size_t num_vertices = 0;

    // Triangulate the mesh, using `get_skinned_vertex()` to fetch each index.
    size_t num_tri_indices = mesh->max_face_triangles * 3;
    uint32_t *tri_indices = calloc(num_tri_indices, sizeof(uint32_t));
    for (size_t face_ix = 0; face_ix < mesh->num_faces; face_ix++) {
        ufbx_face face = mesh->faces.data[face_ix];
        uint32_t num_tris = ufbx_triangulate_face(tri_indices, num_tri_indices, mesh, face);
        for (size_t i = 0; i < num_tris * 3; i++) {
            uint32_t index = tri_indices[i];
            vertices[num_vertices++] = get_skinned_vertex(mesh, skin, index);
        }
    }
    free(tri_indices);
    assert(num_vertices == num_triangles * 3);

    // Create bone descriptions
    size_t num_bones = skin->clusters.count;
    Bone *bones = calloc(num_bones, sizeof(Bone));
    for (size_t i = 0; i < num_bones; i++) {
        ufbx_skin_cluster *cluster = skin->clusters.data[i];
        bones[i].node_index = cluster->bone_node->typed_id;
        bones[i].geometry_to_bone = cluster->geometry_to_bone;
    }

    Mesh result = {
        .bones = bones,
        .num_bones = num_bones,
        .vertices = vertices,
        .num_vertices = num_vertices,
    };
    return result;
}

void matrix_add(ufbx_matrix *dst, const ufbx_matrix *src, float weight)
{
    // `ufbx_matrix` structs are affine 3x4 matrices
    for (size_t i = 0; i < 3*4; i++) {
        dst->v[i] += src->v[i] * weight;
    }
}

// Example implementation that dumps the skinned mesh as an .obj file
void dump_mesh(Mesh mesh, ufbx_scene *scene)
{
    // Resolve skinning (mesh geometry to world matrices)
    ufbx_matrix *geometry_to_world = calloc(mesh.num_bones, sizeof(ufbx_matrix));

    for (size_t i = 0; i < mesh.num_bones; i++) {
        Bone bone = mesh.bones[i];

        // This would really come from some custom scene graph or animation
        ufbx_node *node = scene->nodes.data[bone.node_index];
        ufbx_matrix bone_to_world = node->node_to_world;

        // Composite bind matrix and current transform
        geometry_to_world[i] = ufbx_matrix_mul(&bone_to_world, &bone.geometry_to_bone);
    }

    // Process each vertex
    for (size_t vertex_ix = 0; vertex_ix < mesh.num_vertices; vertex_ix++) {
        Vertex vertex = mesh.vertices[vertex_ix];

        // Weighted sum of the influencing bone matrices
        ufbx_matrix vertex_to_world = { 0 };
        for (size_t i = 0; i < MAX_WEIGHTS; i++) {
            uint32_t bone_ix = vertex.bones[i];
            float weight = vertex.weights[i];
            matrix_add(&vertex_to_world, &geometry_to_world[bone_ix], weight);
        }

        ufbx_vec3 p = ufbx_transform_position(&vertex_to_world, vertex.position);
        printf("v %.3f %.3f %.3f\n", p.x, p.y, p.z);
    }

    // Just a flat list of non-indexed triangle faces
    for (size_t i = 0; i < mesh.num_vertices; i += 3) {
        printf("f %zu %zu %zu\n", i + 1, i + 2, i + 3);
    }

    free(geometry_to_world);
}

int main(int argc, char **argv)
{
    ufbx_scene *scene = ufbx_load_file(argv[1], NULL, NULL);
    assert(scene);

    ufbx_node *node = ufbx_find_node(scene, argv[2]);
    assert(node && node->mesh);

    ufbx_mesh *mesh = node->mesh;
    assert(mesh->skin_deformers.count > 0);
    ufbx_skin_deformer *skin = mesh->skin_deformers.data[0];

    Mesh result = process_skinned_mesh(mesh, skin);
    dump_mesh(result, scene);

    return 0;
}

File: tests/example-frames/deformers/skin-weights.rs
// $dep skinned_human.fbx
// $ cargo build
// $ cargo run --quiet -- skinned_human.fbx Female > result.obj

// -- EXAMPLE_SOURCE --

struct Bone {
    node_index: u32,
    geometry_to_bone: ufbx::Matrix,
}

// Non-indexed triangulated skinned mesh
struct Mesh {
    bones: Vec<Bone>,
    vertices: Vec<Vertex>,
} 

fn process_skinned_mesh(mesh: &ufbx::Mesh, skin: &ufbx::SkinDeformer) -> Mesh {
    let mut vertices: Vec<Vertex> = Vec::new();

    // Triangulate the mesh, using `get_skinned_vertex()` to fetch each index.
    let mut tri_indices = vec![0u32; mesh.max_face_triangles * 3];
    for face_index in 0..mesh.num_faces {
        let face = mesh.faces[face_index];

        // Triangulate the face into `tri_indices[]`.
        let num_tris = mesh.triangulate_face(&mut tri_indices, face);
        let num_tri_corners = (num_tris * 3) as usize;
        for &index in &tri_indices[..num_tri_corners] {
            vertices.push(get_skinned_vertex(mesh, skin, index as usize));
        }
    }
    assert!(vertices.len() == mesh.num_triangles * 3);

    // Create bone descriptions
    let bones = skin.clusters
        .iter()
        .map(|cluster| Bone{
            node_index: cluster.bone_node.as_ref().unwrap().element.typed_id,
            geometry_to_bone: cluster.geometry_to_bone,
        })
        .collect();

    Mesh{ bones, vertices }
}

fn matrix_add(dst: &mut ufbx::Matrix, src: &ufbx::Matrix, weight: f32) {
    // Rust bindings currently have no way of accessing columns/values
    // ergonomically..
    let weight = weight as f64;
    dst.m00 += src.m00 * weight;
    dst.m10 += src.m10 * weight;
    dst.m20 += src.m20 * weight;
    dst.m01 += src.m01 * weight;
    dst.m11 += src.m11 * weight;
    dst.m21 += src.m21 * weight;
    dst.m02 += src.m02 * weight;
    dst.m12 += src.m12 * weight;
    dst.m22 += src.m22 * weight;
    dst.m03 += src.m03 * weight;
    dst.m13 += src.m13 * weight;
    dst.m23 += src.m23 * weight;
}

// Example implementation that dumps the skinned mesh as an .obj file
fn dump_mesh(mesh: &Mesh, scene: &ufbx::Scene) {
    // Resolve skinning (mesh geometry to world matrices)
    let geometry_to_world: Vec<ufbx::Matrix> = mesh.bones
        .iter()
        .map(|bone| {
            let node = &scene.nodes[bone.node_index as usize];
            ufbx::matrix_mul(&node.node_to_world, &bone.geometry_to_bone)
        })
        .collect();

    // Process each vertex
    for vertex in &mesh.vertices {
        // Weighted sum of the influencing bone matrices
        // Rust inherits zero default from C, not identity as default
        let mut vertex_to_world = ufbx::Matrix::default();
        for i in 0..MAX_WEIGHTS {
            let bone_ix = vertex.bones[i] as usize;
            let weight = vertex.weights[i];
            matrix_add(&mut vertex_to_world, &geometry_to_world[bone_ix], weight);
        }

        let p = ufbx::transform_position(&vertex_to_world, vertex.position);
        println!("v {:.3} {:.3} {:.3}", p.x, p.y, p.z);
    }

    // Just a flat list of non-indexed triangle faces
    for tri in 0 .. mesh.vertices.len() / 3 {
        let i = tri * 3;
        println!("f {} {} {}", i + 1, i + 2, i + 3);
    }
}

fn main() {
    let path = std::env::args().nth(1).expect("expected .fbx path");
    let node_name = std::env::args().nth(2).expect("expected node name");

    let scene = ufbx::load_file(&path, Default::default()).expect("failed to load scene");
    let node = scene.find_node(&node_name).expect("could not find node");
    let mesh = node.mesh.as_ref().expect("node did not contain a mesh");
    let skin = mesh.skin_deformers.get(0).expect("no skin deformer in mesh");

    let result = process_skinned_mesh(mesh, skin);
    dump_mesh(&result, &scene);
}
